src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp

```diff
@@ -442,17 +448,82 @@ REDHAWK_PALEXPORT void REDHAWK_PALAPI PalRestoreContext(CONTEXT * pCtx)
 
 static PalHijackCallback g_pHijackCallback;
 
+#ifdef FEATURE_SPECIAL_USER_MODE_APC
+typedef BOOL (WINAPI* QueueUserAPC2Proc)(PAPCFUNC ApcRoutine, HANDLE Thread, ULONG_PTR Data, QUEUE_USER_APC_FLAGS Flags);
+static QueueUserAPC2Proc pfnQueueUserAPC2Proc;
+
+static const QUEUE_USER_APC_FLAGS SpecialUserModeApcWithContextFlags = (QUEUE_USER_APC_FLAGS)
+                                    (QUEUE_USER_APC_FLAGS_SPECIAL_USER_APC |
+                                    QUEUE_USER_APC_CALLBACK_DATA_CONTEXT);
+
+static void NTAPI EmptyActivationHandler(ULONG_PTR parameter){}
+
+static void NTAPI ActivationHandler(ULONG_PTR parameter)
+{
+    APC_CALLBACK_DATA* data = (APC_CALLBACK_DATA*)parameter;
+    g_pHijackCallback(data->ContextRecord, NULL);
+
+    Thread* pThread = (Thread*)data->Parameter;
+    pThread->SetActivationPending(false);
+}
+#endif
+
 REDHAWK_PALEXPORT UInt32_BOOL REDHAWK_PALAPI PalRegisterHijackCallback(_In_ PalHijackCallback callback)
 {
     ASSERT(g_pHijackCallback == NULL);
     g_pHijackCallback = callback;
 
+#ifdef FEATURE_SPECIAL_USER_MODE_APC
+    // Check if we have QueueUserAPC2
+    HMODULE hKernel32 = LoadKernel32dll();
+    pfnQueueUserAPC2Proc = (QueueUserAPC2Proc)GetProcAddress(hKernel32, "QueueUserAPC2");
+
+    // See if QueueUserAPC2 supports the special user-mode APC with a callback that includes the interrupted CONTEXT
+    if (pfnQueueUserAPC2Proc != NULL)
+    {
+        if (!(*pfnQueueUserAPC2Proc)(EmptyActivationHandler, GetCurrentThread(), 0, SpecialUserModeApcWithContextFlags))
```

Comment:
On Win10 this will always fail, but the API is present. We would need to prevent trying over and over. 
With some sentinel value for "unknown" state and NULL for "does not work" this might be possible.

Reply:
Right, we can have a bool flag that says the API failed with "SpecialUserModeApcWithContextFlags not supported error". Once the flag gets set, we will stop trying.

Reply:
I've uses a sentinel value for uninitialized state. Hijacking pass is single-threaded, so initializing on demand is simple. 
Maybe will save us some ticks at startup.